### ISSUE TITLE
Make Polymer WebJars Work

### ISSUE THREAD

**Issue by jamesward** [webjars/webjars#1670]
Which repos are problematic?

What do we do about https://github.com/Polymer/polymer/issues/4912 ?

What do we do about artifact naming?  GitHub-based?  Lowercase?

**Comment by jamesward** [comment]
One possible solution is to have a special group id for handling this.  If we are just talking about https://github.com/PolymerElements/ then maybe we have `org.webjars.polymerelements` and have artifacts deployed based on repo names like `iron-elements`.  One possible complication is if any of these packages have dependencies outside of PolymerElements.  Also in this package we could flatten the dir structure so that the cross-package paths work correctly.

**Comment by jamesward** [comment]
After further thought and discussion with @Legioth I think we should do a new GitHub WebJar type that will have the following traits:

- GroupId = `org.webjars.github.[ORG]`
- ArtifactId = `[REPO]`
- GroupId and ArtifactId will be lowercased
- Versions will always be a git tag with any proceeding `v` stripped off. (possibly add commit id versions in the future)
- Metadata will be read from `bower.json` with a fallback to `package.json`
- License will be read from github-license-service (ignoring the Bower and NPM info since it is often wrong / hard to determine
- Dependency info will always be converted to GitHub WebJars.  For instance, if the `bower.json` has a dependency on the `iron-elements` Bower package, it will be converted to a dependency on `org.webjars.github.polymerelements:iron-elements`.
- Consider flattening the dir structure to `META-INF/resources/[contents]`

The new WebJar type will be exposed on the webjars.org site in the deploy, search, and listing screens.

**Comment by Legioth** [comment]
How are dependencies converted to GitHub WebJars for npm packages? `package.json` can have a `repository` field, but there's no guarantee that it will point to GitHub and not e.g. Bitbucket or even some custom `svn` URL. This is on the other hand really straightforward for Bower dependencies. 

There's also the possibility for the weird case of the version defined in `package.json` not being in sync with any tag on GitHub, which means that it's not 100% safe to directly convert from an npm dependency version to a GitHub tag. One possible approach here would be to always use the version number defined in `package.json` as the ArtifactId.

Could it be that npm packages are already best handled through the current mechanism? I'm not too familiar with how current WebJars work in practice for those, but I would at least imagine that there's no problem with multiple names for the same logical package?

For the directory structure, I would suggest `META-INF/resources/webjars/[native package name]/[contents]`.
- Keep `webjars` as a prefix to reduce the impact on the URL namespace available for the application's own content. Could also consider using `webjars/bower_components` and `webjars/node_modules` to exactly match the directory structure that the packages expect to be in.
- Getting rid of the version number will help simplify things in most cases, even though it would rule out the theoretical possibility of having multiple versions of the same library available.
- Use the actual Bower or npm package name instead of the WebJar name (i.e. the ArtifactId). Those two are usually the same, but not necessarily in cases when a project has been forked. This is helpful since the native name defines the target directory used by `bower install` or `npm install`.

**Comment by jamesward** [comment]
You are right that NPM deps which are not on GitHub would not be able to be converted.  I think I'm willing to live with that limitation.

I think we can just ignore the version in the metadata and only use the git tag.  See any issue with that?

The NPM WebJar type won't go away so people can continue using that.

Yes, I like your proposed dir structure.  Since Java package managers do not allow multiple versions of the same lib, I think it is safe to omit the version from the dir structure.  The main reason why I put it in there was to help with CDN caching.  For the `[native package name` I think you are right that we should actually use the name specified in the metadata instead of the GitHub repo name.

**Comment by Legioth** [comment]
I analyzed a completely unrepresentative sample of 1309 `package.json` files based on doing `npm install polymer-cli`. This sample didn't contain any packages with a non-GitHub repository, but it included 34 packages without any repository information at all: `find . -name package.json | xargs jq .repository.url | grep -v github | wc -l`.

Once you encounter one transitive dependency that doesn't have a GitHub repository, you would have to include those as `org.webjars.nmp` dependencies instead. This would introduce the problem of having two variants of any shared transitive dependencies.

Regarding versions, the issue is that some module might depend on a specific version of another module, but the developer forgot to push a corresponding tag to GitHub even though the version has been published to the npm directory. Even worse is if the GitHub tag is accidentally based on a different commit compared to the npm directory version.

The end result of both issues is that using `org.webjars.github` is a gamble that might explode at any time, whereas `org.webjars.npm` would be a completely solution in basically any case.

With that said, I think that using WebJars with Polymer 3 would be significantly simplified if there would be npm WebJars with the same directory structure as suggested here, since that would again match the assumptions that modules have to make about dependencies to other modules. In that way, we wouldn't have to rely on the WebJars Locator to translate between `/[WebJar name]/[version]/` and `/[native package name]/`.

**Comment by jamesward** [comment]
Because the NPM package without a repo url thing was continuing to be a problem I created an admin function that allows me to manually specify these.  While not ideal it has helped to deal with various missing deps.

We've run into tags not existing but have had pretty good response to filing issues for authors to push tags.  In the case of Bower, does it always use git tags when resolving deps?  I think you are right that NPM goes the other way and always uses the NPM metadata.

I don't think we can change the dir structure of `org.webjars.npm` so we'd still have to use the servlet filter workaround there.  I think there might also be risks with changing it for a new WebJar  type.  So I need to think more about that.

**Comment by Legioth** [comment]
I'm not sure I understand how things would work in practice if someone tries to create an `org.webjars.github` WebJar for an npm package with a dependency without repository information. What would the GroupId and ArtifactId look like for a dependency without a GitHub URL?

Bower versions are indeed always based on SCM tags (not necessarily GitHub in that case either, but it seems to be less common to use anything else). Some packages have a `verison` field in `bower.json`, but it has been deprecated and is completely ignored. The only centralized part of Bower is the registry that converts shorthand names into SCM coordinates.

npm on the other hand is more similar to e.g. Maven in the sense that version numbers and package contents are based on what's pushed to a centralized directory which might or might not be in sync with what's pushed to a SCM repository.

One additional thing that came to mind about the directory structure is that in most cases, including the version number for cache management reasons is quite redundant. The reason for this is that for a highly performant production setup, it is also necessary to bundle the original resources. One of the promises of HTTP/2 was that it would not longer be necessary to bundle dependencies. The actual result is, however, that it's it's nowadays acceptable to load tens of dependencies, but hundeds will still cause significant overhead. Polymer encourages splitting things up to quite small pieces, so it's not uncommon to have over 200 different files to load if not using bundling.

**Comment by jamesward** [comment]
> What would the GroupId and ArtifactId look like for a dependency without a GitHub URL?

I guess I'm thinking that it is common in cases where the repository isn't specified that it actually does exist on GitHub.  But I think you are right that the exceptions to this will be a real pain.  So thinking further about it, I think doing NPM packages like this won't work out well.

> including the version number for cache management reasons is quite redundant.

The main reason for this is that far future expires for URLs with artifact versions is super easy.  You can basically tell the browser that an artifact with the versioned URL will never change.

**Comment by jamesward** [comment]
Based on the potential issues with converting NPM dependencies correctly I'd like to update my proposal to only be for Bower artifacts:

- GroupId = `org.webjars.bower.[ORG]`
- ArtifactId = `[REPO NAME]`
- GroupId and ArtifactId will be lowercased
- Versions will always be a git tag with any proceeding v stripped off. (possibly add commit id versions in the future)
- Metadata will be read from `bower.json`
- License will try to be read from the Bower info, with a fallback to the github-license-service
- Dependency info will always be converted to this form. For instance, if the `bower.json` has a dependency on the `iron-elements` Bower package, it will be converted to a dependency on `org.webjars.bower.polymerelements:iron-elements`.
- Flatten the dir structure to `META-INF/resources/webjars/[ARTIFACT ID]/[CONTENTS]`

**Comment by Legioth** [comment]
Was my suggestion about using `[BOWER NAME]` instead of `[ARTIFACT ID]` left out by accident or for some specific reason? Other than that, I cannot anticipate any problems with the latest proposal.

**Comment by jamesward** [comment]
Whoops.  Just a mistake.  Good catch!  Note that there will be a potential mismatch between the Artifact ID and the dir name.  Which I think is ok.  I think webjars-locator requires these be the same but without the version, we should need to use webjars-locator... Actually Play uses it to extract the contents of the WebJars which might require us to fix something there.

**Comment by SomeoneToIgnore** [comment]
Hey, so, any actions considered yet?

I have two pull requests that look like a good start: 

https://github.com/webjars/webjars/pull/1669
https://github.com/webjars/webjars/pull/1667

**Comment by jamesward** [comment]
I almost have a branch ready that will at least have all the tests for this.  But there has been a bunch of refactoring to get there.  Ultimately we need a new WebJar type that implements these changes.

**Comment by SomeoneToIgnore** [comment]
Oh, I see. If I can be of any help, let me know.
Will be waiting for this then.

**Comment by jamesward** [comment]
Here is the WIP branch: https://github.com/webjars/webjars/tree/wip-bower-new

Since it looks like you all are working with the WebJar extractor stuff I think we should consider keeping the dir structure the same as the other WebJar types:
`META-INF/resources/webjars/[ARTIFACT ID]/[VERSION]/[CONTENTS]`
This would alleviate potential issues with the CDN setup and possibly other unforeseen issues.  But let me know if you think that is an option.

**Comment by Legioth** [comment]
We'd actually like to get rid of the `webjars-locator` dependency since it requires different protocol handlers for different servers, which in turn means that we cannot provide a set of dependencies that works out of the box in all cases.

**Comment by SomeoneToIgnore** [comment]
Finally had managed to have a look at the branch, nice new spec you've got there.

To me it also looks more natural to have `META-INF/resources/webjars/[BOWER NAME]/[CONTENTS]` as the new WebJar structure: if we have similar structure for the npm packages, `WebJarExtractor` code can be simplified a lot.
Was there any reason for including it into previous versions? From what I can think of, it does not bring any value, since we cannot have multiple jars different by version number only.

One more question related to the future plans on the old bower WebJars: would both bower WebJar versions supported in the future?
To me it looks like we should drop the old ones after some point and a migration plan would be an awesome thing to have.

**Comment by jamesward** [comment]
The main reason for including the version in the path is that it makes it trivial to put far-future expires headers on the resources.  So once the browser has `/webjars/foo/1.0.0/foo.js` it never has to ask for it again.  Without the version you can use etags but those still require a request to see if it is the newest.

Out of the gate we will have a new WebJar type (currently thinking about calling it Bower GitHub) along side the old one.  We could eventually drop support for deploying the old Bower type.  But I haven't thought that through yet.

**Comment by Legioth** [comment]
JS modules designed for use with bower expect to find their transitive dependencies from `../[bower name]/foo.js`, which breaks down if the actual location is `../../[webjar name]/[version]/foo.js`.

Many JS modern JS libraries are spread out over multiple smaller files instead of one big file with the entire library. This means that any site or app that actually bothers to optimize for performance will be bundling the files together using e.g. webpack or Polymer Build for production builds and only use the original files during development.

**Comment by jamesward** [comment]
One of the goals of WebJars is to be able to be served over HTTP.  So I want to make sure we support that use case well.

Although, I still think it's an anti-pattern to assume a cross-module dir structure but I don't think I'm going to fix that.  :)

I'm totally willing to try without the version in there.  And I think I can still support the far-future expires thing with some additional support in Play, a Servlet, etc.  So let's stick with the versionless path on this one.

**Comment by jamesward** [comment]
I have all the tests passing on the new branch.  But I'd like someone to sanity check this and make sure the tests are covering the functionality we decided on.  Changes at:
https://github.com/webjars/webjars/compare/wip-bower-new

**Comment by SomeoneToIgnore** [comment]
I'm not fluent with Scala, but logically it looks fine to me.

I have a few notes: 

* https://github.com/webjars/webjars/compare/wip-bower-new#diff-931c0fcdd6c6c0dae3841b9d99070c34R23 looks like `groupdId` can be inlined into `groupIdQuery`, same as done for `Bower` and  `BowerGitHub` implementations.

* `lookup` method defined in `Bower` used in `BowerGitHub` only. Should we move it to that class instead? Can also make it private, I guess.

* Should we add `org.webjars.bowergithub` to a pathPrefix in
https://github.com/webjars/webjars/compare/wip-bower-new#diff-8e2d27bf629e02be090dda3d0e2a02e4R67 ?
It looks like we should, since we have `groupId` method that adds this prefix.
I also propose to extract this prefix into some variable to eliminate the duplication

Otherwise it looks nice :+1:

**Comment by Artur-** [comment]
I looked through the tests and they make sense and cover what I can think of.

Does this resolve #1664 also (make Bower ignore work consistently regardless of how the URL is given)?

**Comment by Artur-** [comment]
Also, would it be possible to get this deployed somewhere for real world validation before going live?

**Comment by SomeoneToIgnore** [comment]
I have also one thing to consider: Maven (and, I presume, Gradle also) have a bit different approach to artifacts versions, described here: https://cwiki.apache.org/confluence/display/MAVENOLD/Dependency+Mediation+and+Conflict+Resolution#DependencyMediationandConflictResolution-DependencyVersionRanges

TL;DR this is not semver. `1.0.alpha3` **does not** belong to the `[1.0, 3.0)` range.
The Maven way of specifying the version is `<major>.<minor>.<revision>([ -<qualififer> ] | [ -<build> ])`, so `1.0-alpha-3` is the right version in this case that will fit in the range.

**Comment by jamesward** [comment]
Thanks all for the feedback!

@SomeoneToIgnore 
 - I collapsed `groupdId` into `groupIdQuery` in `NPM`
 - I'd like to leave `Bower.lookup` where it is since it is really about the Bower repo, so not specific to `BowerGitHub`
 - For `BowerGitHub` I actually can't put actual the `groupId` into the `pathPrefix` because that'd break the ability for relative module loading (like is used in Polymer).  We could put `org.webjars.bowergithub` in the path prefix.  But having a mismatch between that part of the `pathPrefix` and the `groupId` might cause issues.
 - For semver to Maven stuff... I wonder if we can get a test going that uses Maven's range resolver.

@Artur- 
 - Yeah, I will deploy this on a test server soon.
 - I will add a test for the ignore thing. It looks like right now I'm just using the `Bower.archive` for `BowerGitHub` which would have the #1664 issue.

**Comment by SomeoneToIgnore** [comment]
Not sure if we can extract the resolution code out of Maven.
We can try IT testing with custom poms built and Maven output analyzed (which is cumbersome though) or testing actual rules, for instance, here's a good description: https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN401

**Comment by jamesward** [comment]
The ignore stuff is now implemented for `BowerGitHub`.  I think the only thing left is potentially changing the SemVer -> Maven converted.  I'll get this deployed on a test server soon so we can do further testing.

**Comment by Legioth** [comment]
Any news on the test server?

**Comment by jamesward** [comment]
The test server is ready!

Try it, like:
```
curl -X POST 'http://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=jquery&version=2.1.1'
```

I've disable Maven Central sync but you can check the contents on BinTray, like:
https://bintray.com/webjars/maven/org.webjars.bowergithub.jquery%3Ajquery-dist

**Comment by SomeoneToIgnore** [comment]
Ok, so I've tried to publish https://github.com/vaadin/vaadin-button/ (v2.0.0-alpha5 version), https://github.com/PolymerElements/iron-meta (v2.0.4 and v2.0.4) and https://github.com/Polymer/polymer (v2.4.0-rc.1) 

Observations:

* `v` prefix in version is mandatory when specifying GitHub links, it is also persisted in the resulting version (artifact, pom, everywhere), this is bad
* Because of the `v` prefix, pom.xml `tag` is spoiled: `<tag>vv2.0.0-alpha5</tag>`
* iron-meta of version `v2.0.4` is not publishing, fails with 
```
<html>
<head><title>405 Not Allowed</title></head>
<body bgcolor="white">
<center><h1>405 Not Allowed</h1></center>
<hr><center>nginx</center>
</body>
</html>
```
have tried 5 times, still fails. Need to understand what's wrong
* iron-meta of version `v2.0.3` was published sucessfully
* Everything is lowercased, which is good.
* Dependencies for vaadin-button pom have the correct versions, awesome
* Dependencies for polymer pom have their versions totally broken because they use `v` prefix in the declaration, looks like we need to pay attention to those also

So, for now the new version looks better, but still have issues with `v` versions.
Also, `iron-meta 2.0.4` deployment looks suspicious.

I'll continue my investigations and see if I can find something else.

**Comment by SomeoneToIgnore** [comment]
Another non-deployable component: `curl -X POST 'http://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=https://github.com/Polydile/dile-valid-date&version=v0.0.5'`

**Comment by jamesward** [comment]
Thanks!  I've cleared out those artifacts with the `v` prefix and fixed that:
https://github.com/webjars/webjars/commit/862d841ebfc3878360ad350a6ae5dcaa13d67e7b

For the vcs tag in the pom, we should fix that but it's not a showstopper since it doesn't really get used anywhere.

I've fixed the `v` + semver range conversion issue:
https://github.com/webjars/webjars/commit/319f4f564f1e76ba66490d8ddf20c2329e541a01

I can't reproduce the `405 Not Allowed` so I think this must have been a transient issue.  Let me know if it persists.

These changes have been deployed so keep on testing. :)

**Comment by Artur-** [comment]
https://dl.bintray.com/webjars/maven/org/webjars/bowergithub/polymer/polymer/2.0.2/polymer-2.0.2.pom has a dependency on 
```
        <dependency>
            <groupId>org.webjars.bowergithub.polymer</groupId>
            <artifactId>webcomponentsjs</artifactId>
            <version>^v1.0.2</version>
        </dependency>
```

**Comment by Artur-** [comment]
curl -X POST "http://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=webcomponentsjs&version=v1.0.19"

Deploys the files into https://dl.bintray.com/webjars/maven/org/webjars/bowergithub/polymer/webcomponentsjs/ but there is no https://github.com/polymer/webcomponentsjs. It has been renamed to https://github.com/webcomponents/webcomponentsjs

**Comment by jamesward** [comment]
I've redeployed the polymer 2.0.2 artifact to BinTray after some additional fixes.

The repo move case needs to be dealt with on the artifact and dependency use cases.  I'll work on that.

**Comment by Artur-** [comment]
Deploy quite a lot of stuff which seems to all work ok but now when trying
`curl -X POST "http://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=https://github.com/PolymerElements/iron-behaviors&version=v2.0.0"`

I only get `405 Not Allowed`. Initially I got `Application Error` a couple of time, then it has been 405 ever since. 

https://dl.bintray.com/webjars/maven/org/webjars/bowergithub/polymerelements/iron-behaviors/2.0.0/ exists but is empty

Deploying version v1.0.18 works fine

**Comment by jamesward** [comment]
Thanks @Artur- for testing this further.  It looks like there is a case where the artifact hasn't been deployed to Maven Central but the version exists on BinTray.  I'll work on a fix.

**Comment by jamesward** [comment]
Now I can't reproduce that `405` error so I think it is a transient issue with BinTray.

**Comment by jamesward** [comment]
I wonder if there is some API rate limiting we hit sometimes...

**Comment by jamesward** [comment]
A fix for moved repos will be deployed in about 15 minutes.  Some more testing would be good.  At this point I'm not aware of any outstanding blockers.

**Comment by Artur-** [comment]
My testing so far has not revealed any more issues. 

There is only the inconvenience that you need to deploy dependencies separately. If they were autodeployed, the process would be much smoother. I would not consider this a blocker in any way

**Comment by jamesward** [comment]
Cool.  For auto transitive dependency deployment, let's track that in #1186.

Here is my PR: #1702

Will give this another day or so for testing.

**Comment by jamesward** [comment]
This has been deployed to prod!  If there are any problems please file new issues. I still need to create the deploy and search UI for the new Bower GitHub type: https://github.com/webjars/webjars/issues/1703

**Comment by Artur-** [comment]
Awesome work! Is the deployment REST API still available?

**Comment by jamesward** [comment]
Yup! That is currently the only way to deploy Bower GitHub artifacts.  Like:
```
curl -X POST 'https://www.webjars.org/deploy?webJarType=BowerGitHub&nameOrUrlish=jquery&version=3.3.1'
```

**Comment by SomeoneToIgnore** [comment]
Awesome, thanks.

When Maven Central deployment would be enabled?

```
$ curl -iX POST 'http://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=vaadin-button&version=2.0.0-alpha5'
HTTP/1.1 500 Internal Server Error
Server: Cowboy
Connection: keep-alive
Date: Tue, 23 Jan 2018 15:06:10 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 36
Via: 1.1 vegur

Deployment to Maven Central Disabled
```

**Comment by jamesward** [comment]
@SomeoneToIgnore Use the prod server instead `www.webjars.org`

**Comment by SomeoneToIgnore** [comment]
Ah, right, thanks.

**Comment by Artur-** [comment]
Now that deployment is asynchronous, is there any way to check what happened with the deployment? E.g. if I do
```
curl -X POST "https://www.webjars.org/deploy?webJarType=BowerGitHub&nameOrUrlish=https://github.com/GeoloeG/gold-password-input.git&version=v2.1.0"
```
nothing ends up in https://dl.bintray.com/webjars/maven/org/webjars/bowergithub/geoloeg but I have no idea why

**Comment by jamesward** [comment]
@Artur- I've deployed a fix that streams the output back.  It still needs some cleanup but should help you get the errors.

**Comment by Artur-** [comment]
Thanks, "Ref v2.1.0 can not be resolved" helps quite a bit to realize that the tag is "2.1.0" and not "v2.1.0" :)
